File: easy/872. Leaf-Similar Trees/872. Leaf-Similar Trees.go
package main

import (
	"reflect"
)

/**
 * Definition for a binary tree node.
 */

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func dfs(root *TreeNode, leafList *[]int) {
	if root != nil {
		if root.Left == nil && root.Right == nil {
			*leafList = append(*leafList, root.Val)
		}
		dfs(root.Left, leafList)
		dfs(root.Right, leafList)
	}
}

func leafSimilar(root1 *TreeNode, root2 *TreeNode) bool {
	list1 := make([]int, 0)
	list2 := make([]int, 0)

	dfs(root1, &list1)
	dfs(root2, &list2)

	return reflect.DeepEqual(list1, list2)
}

File: easy/872. Leaf-Similar Trees/872. Leaf-Similar Trees.py
from typing import List, Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        def dfs(root: Optional[TreeNode], leaf_list: List[int]) -> None:
            if root:
                if not root.left and not root.right:
                    leaf_list.append(root.val)
                dfs(root.left, leaf_list)
                dfs(root.right, leaf_list)

        list1 = []
        list2 = []

        dfs(root1, list1)
        dfs(root2, list2)

        return list1 == list2
